File: tests/test_squash.py
"""Tests for FS.squash()."""
import pytest
from vost import GitStore


@pytest.fixture
def store(tmp_path):
    return GitStore.open(str(tmp_path / "test.git"), branch="main")


def test_squash_creates_root_commit(store):
    fs = store.branches["main"].write("a.txt", b"a")
    fs = fs.write("b.txt", b"b")  # 3 commits total

    squashed = fs.squash()
    assert squashed.read("a.txt") == b"a"
    assert squashed.read("b.txt") == b"b"
    assert squashed.parent is None  # root commit
    assert squashed.writable is False


def test_squash_preserves_tree_hash(store):
    fs = store.branches["main"].write("data.txt", b"hello")
    squashed = fs.squash()
    assert squashed.tree_hash == fs.tree_hash


def test_squash_with_parent(store):
    tip = store.branches["main"]
    fs = tip.write("a.txt", b"a")
    fs = fs.write("b.txt", b"b")

    squashed = fs.squash(parent=tip)
    assert squashed.parent is not None
    assert squashed.parent.commit_hash == tip.commit_hash
    assert squashed.read("b.txt") == b"b"


def test_squash_with_message(store):
    fs = store.branches["main"].write("a.txt", b"a")
    squashed = fs.squash(message="Custom squash message")
    assert squashed.message.startswith("Custom squash message")


def test_squash_default_message(store):
    fs = store.branches["main"].write("a.txt", b"a")
    squashed = fs.squash()
    assert "squash" in squashed.message.lower()


def test_squash_assign_to_branch(store):
    fs = store.branches["main"].write("a.txt", b"a")
    fs = fs.write("b.txt", b"b")

    squashed = fs.squash()
    store.branches["squashed"] = squashed

    result = store.branches["squashed"]
    assert result.read("a.txt") == b"a"
    assert result.parent is None


def test_squash_in_place(store):
    """Squash a branch in place."""
    fs = store.branches["main"].write("a.txt", b"a")
    fs = fs.write("b.txt", b"b")

    squashed = fs.squash()
    store.branches["main"] = squashed

    result = store.branches["main"]
    assert result.read("b.txt") == b"b"
    assert result.parent is None

File: cpp/tests/test_squash.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <filesystem>
#include <string>
#include <thread>
#include <chrono>

namespace fs = std::filesystem;

static fs::path make_temp_repo() {
    auto tmp = fs::temp_directory_path() /
               ("vost_sqtest_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_store(const fs::path& path,
                                  const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

TEST_CASE("Squash: creates root commit (no parents)", "[squash]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "hello");
    snap = snap.write_text("b.txt", "world");

    auto squashed = snap.squash();

    // The squashed commit should exist
    REQUIRE(squashed.commit_hash().has_value());
    // It should differ from the original commit
    CHECK(squashed.commit_hash() != snap.commit_hash());
    // It should have no parent (root commit)
    CHECK_FALSE(squashed.parent().has_value());
    // It should be read-only (detached)
    CHECK_FALSE(squashed.writable());
    // No ref name
    CHECK_FALSE(squashed.ref_name().has_value());

    fs::remove_all(path);
}

TEST_CASE("Squash: preserves tree hash", "[squash]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "hello");
    snap = snap.write_text("b.txt", "world");

    auto squashed = snap.squash();

    CHECK(squashed.tree_hash() == snap.tree_hash());
    // Verify content is readable
    CHECK(squashed.read_text("a.txt") == "hello");
    CHECK(squashed.read_text("b.txt") == "world");

    fs::remove_all(path);
}

TEST_CASE("Squash: with parent", "[squash]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "v1");
    auto parent_snap = snap.squash(); // root squash as parent

    snap = snap.write_text("a.txt", "v2");
    auto squashed = snap.squash(parent_snap);

    // Should have a parent
    REQUIRE(squashed.parent().has_value());
    CHECK(squashed.parent()->commit_hash() == parent_snap.commit_hash());
    // Tree matches source
    CHECK(squashed.tree_hash() == snap.tree_hash());
    CHECK(squashed.read_text("a.txt") == "v2");

    fs::remove_all(path);
}

TEST_CASE("Squash: custom message", "[squash]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "hello");
    auto squashed = snap.squash(std::nullopt, "my custom message");

    CHECK(squashed.message() == "my custom message");

    fs::remove_all(path);
}

TEST_CASE("Squash: assign to branch", "[squash]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "hello");
    snap = snap.write_text("b.txt", "world");

    auto squashed = snap.squash();

    // Assign the squashed commit to a new branch
    store.branches().set("squashed", squashed);
    auto branch_snap = store.branches().get("squashed");

    CHECK(branch_snap.tree_hash() == snap.tree_hash());
    CHECK(branch_snap.read_text("a.txt") == "hello");
    CHECK(branch_snap.read_text("b.txt") == "world");
    // The branch snapshot should have no parent (root commit)
    CHECK_FALSE(branch_snap.parent().has_value());

    fs::remove_all(path);
}
